{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import geonomics as gnx\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import scipy.fftpack\n",
    "from nlmpy import nlmpy\n",
    "from pykrige.ok import OrdinaryKriging"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def DataFrameGEA(model, spp = 0):\n",
    "    \"\"\"Makes a dataframe for GEA analysis with genotypes, coordinates, and env values for all individuals.\n",
    "    \n",
    "    Keyword arguments:\n",
    "    model -- gnx model\n",
    "    spp -- species index (default: 0)\n",
    "    \n",
    "    Output:\n",
    "    Dataframe (Columns are loci and rows are individuals)\n",
    "    \"\"\"    \n",
    "    #get genotypes\n",
    "    gen = model.comm[spp]._get_genotypes()\n",
    "    \n",
    "    #loop to convert binary allele genotypes (0|1) into single digit genotypes based on allele count (0, 0.5, 1)\n",
    "    geno_ind = [] \n",
    "    for ind in range(len(gen)):\n",
    "        geno_loc = []\n",
    "        for loci in range(len(gen[ind])):\n",
    "            genotype = gen[ind][loci].mean() #Codes as 0/0.5/1\n",
    "            geno_loc.append(genotype)\n",
    "        geno_ind.append(geno_loc)   \n",
    "    \n",
    "    #convert to dataframe\n",
    "    gea_df = pd.DataFrame(geno_ind)\n",
    "    \n",
    "    #get environmental data\n",
    "    env = model.comm[spp]._get_e()\n",
    "    gea_df['env0'] = list(env[:,1]) \n",
    "    gea_df['env1'] = list(env[:,2]) \n",
    "\n",
    "    #get coords\n",
    "    coord = model.comm[spp]._get_coords()\n",
    "    gea_df['x'] = list(coord[:,0])\n",
    "    gea_df['y'] = list(coord[:,1])\n",
    "\n",
    "    return gea_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_unif_array(n):\n",
    "    \"\"\"Makes a square array of ones, size n x n cells.\"\"\"\n",
    "    array = np.ones((n,n))\n",
    "    return array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "env0 = np.genfromtxt('data/seed1_env1_H05.csv',delimiter=',')\n",
    "env1 = np.genfromtxt('data/seed1_env2_H05_r06.csv',delimiter=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(42)\n",
    "# selection_demo.py\n",
    "\n",
    "# This is a parameters file generated by Geonomics\n",
    "# (by the gnx.make_parameters_file() function).\n",
    "\n",
    "\n",
    "                   #   ::::::          :::    :: :::::::::::#\n",
    "             #::::::    ::::   :::      ::    :: :: ::::::::::: ::#\n",
    "          #:::::::::     ::            ::   :::::::::::::::::::::::::#\n",
    "        #::::::::::                      :::::::::: :::::: ::::::::  ::#\n",
    "      #  : ::::  ::                    ::::  : ::    :::::::: : ::  :    #\n",
    "     # GGGGG :EEEE: OOOOO   NN   NN   OOOOO   MM   MM IIIIII  CCCCC SSSSS #\n",
    "    # GG     EE    OO   OO  NNN  NN  OO   OO  MM   MM   II   CC     SS     #\n",
    "    # GG     EE   OO     OO NN N NN OO     OO MMM MMM   II   CC     SSSSSS #\n",
    "    # GG GGG EEEE OO     OO NN  NNN OO     OO MM M MM   II   CC         SS #\n",
    "    # GG   G EE    OO   OO  NN   NN  OO   OO  MM   MM   II   CC        SSS #\n",
    "     # GGGGG :EEEE: OOOOO   NN   NN   OOOOO   MM   MM IIIIII  CCCCC SSSSS #\n",
    "      #    : ::::::::               :::::::::: ::              ::  :   : #\n",
    "        #:    :::::                    :::::: :::             :::::::  #\n",
    "          #    :::                      :::::  ::              ::::: #\n",
    "             #  ::                      ::::                      #\n",
    "                   #                                        #\n",
    "                      #  :: ::    :::             #\n",
    "\n",
    "\n",
    "params = {\n",
    "#--------------------------------------------------------------------------#\n",
    "\n",
    "#-----------------#\n",
    "#--- LANDSCAPE ---#\n",
    "#-----------------#\n",
    "    'landscape': {\n",
    "\n",
    "    #------------#\n",
    "    #--- main ---#\n",
    "    #------------#\n",
    "        'main': {\n",
    "            #x,y (a.k.a. j,i) dimensions of the Landscape\n",
    "            'dim':                      (40,40),\n",
    "            #x,y resolution of the Landscape\n",
    "            'res':                      (1,1),\n",
    "            #x,y coords of upper-left corner of the Landscape\n",
    "            'ulc':                      (0,0),\n",
    "            #projection of the Landscape\n",
    "            'prj':                      None,\n",
    "            }, # <END> 'main'\n",
    "\n",
    "    #--------------#\n",
    "    #--- layers ---#\n",
    "    #--------------#\n",
    "        'layers': {\n",
    "\n",
    "            #layer name (LAYER NAMES MUST BE UNIQUE!)\n",
    "            'lyr_0': {\n",
    "\n",
    "        #-------------------------------------#\n",
    "        #--- layer num. 0: init parameters ---#\n",
    "        #-------------------------------------#\n",
    "\n",
    "                #initiating parameters for this layer\n",
    "                'init': {\n",
    "\n",
    "                    #parameters for a 'defined'-type Layer\n",
    "                    'defined': {\n",
    "                        #raster to use for the Layer\n",
    "                        'rast':                   make_unif_array(40),\n",
    "                        #point coordinates\n",
    "                        'pts':                    None,\n",
    "                        #point values\n",
    "                        'vals':                   None,\n",
    "                        #interpolation method {None, 'linear', 'cubic',\n",
    "                        #'nearest'}\n",
    "                        'interp_method':          None,\n",
    "\n",
    "                        }, # <END> 'defined'\n",
    "\n",
    "                    }, # <END> 'init'\n",
    "\n",
    "                }, # <END> layer num. 0\n",
    "\n",
    "\n",
    "            #layer name (LAYER NAMES MUST BE UNIQUE!)\n",
    "            'lyr_1': {\n",
    "\n",
    "        #-------------------------------------#\n",
    "        #--- layer num. 1: init parameters ---#\n",
    "        #-------------------------------------#\n",
    "\n",
    "                #initiating parameters for this layer\n",
    "                'init': {\n",
    "\n",
    "                    #parameters for a 'defined'-type Layer\n",
    "                    'defined': {\n",
    "                        #raster to use for the Layer\n",
    "                        'rast':                   env1,\n",
    "                        #point coordinates\n",
    "                        'pts':                    None,\n",
    "                        #point values\n",
    "                        'vals':                   None,\n",
    "                        #interpolation method {None, 'linear', 'cubic',\n",
    "                        #'nearest'}\n",
    "                        'interp_method':          None,\n",
    "\n",
    "                        }, # <END> 'defined'\n",
    "\n",
    "                    }, # <END> 'init'\n",
    "\n",
    "                }, # <END> layer num. 1\n",
    "            \n",
    "            #layer name (LAYER NAMES MUST BE UNIQUE!)\n",
    "            'lyr_2': {\n",
    "        #-------------------------------------#\n",
    "        #--- layer num. 2: init parameters ---#\n",
    "        #-------------------------------------#\n",
    "\n",
    "                #initiating parameters for this layer\n",
    "                'init': {\n",
    "\n",
    "                    #parameters for a 'defined'-type Layer\n",
    "                    'defined': {\n",
    "                        #raster to use for the Layer\n",
    "                        'rast':                   env2,\n",
    "                        #point coordinates\n",
    "                        'pts':                    None,\n",
    "                        #point values\n",
    "                        'vals':                   None,\n",
    "                        #interpolation method {None, 'linear', 'cubic',\n",
    "                        #'nearest'}\n",
    "                        'interp_method':          None,\n",
    "\n",
    "                        }, # <END> 'defined'\n",
    "\n",
    "                    }, # <END> 'init'\n",
    "\n",
    "                }, # <END> layer num. 2\n",
    "\n",
    "\n",
    "    #### NOTE: Individual Layers' sections can be copy-and-pasted (and\n",
    "    #### assigned distinct keys and names), to create additional Layers.\n",
    "\n",
    "\n",
    "            } # <END> 'layers'\n",
    "\n",
    "        }, # <END> 'landscape'\n",
    "\n",
    "\n",
    "#-------------------------------------------------------------------------#\n",
    "    \n",
    "#-----------------#\n",
    "#--- COMMUNITY ---#\n",
    "#-----------------# \n",
    "    'comm': {\n",
    "\n",
    "        'species': {\n",
    "\n",
    "            #species name (SPECIES NAMES MUST BE UNIQUE!)\n",
    "            'spp_0': {\n",
    "\n",
    "            #-----------------------------------#\n",
    "            #--- spp num. 0: init parameters ---#\n",
    "            #-----------------------------------#\n",
    "\n",
    "                'init': {\n",
    "                    #starting number of individs\n",
    "                    'N':                250,\n",
    "                    #carrying-capacity Layer name\n",
    "                    'K_layer':          'lyr_0',\n",
    "                    \n",
    "###########################################################################\n",
    "# CONSIDER CHANGING THIS PARAMETER                                        #\n",
    "                    #multiplicative factor for carrying-capacity layer    #\n",
    "                    'K_factor':         5,                                #\n",
    "###########################################################################\n",
    "                    \n",
    "                    }, # <END> 'init'\n",
    "\n",
    "            #-------------------------------------#\n",
    "            #--- spp num. 0: mating parameters ---#\n",
    "            #-------------------------------------#\n",
    "\n",
    "                'mating'    : {\n",
    "                    #age(s) at sexual maturity (if tuple, female first)\n",
    "                    'repro_age':                0,\n",
    "                    #whether to assign sexes\n",
    "                    'sex':                      False,\n",
    "                    #ratio of males to females\n",
    "                    'sex_ratio':                1/1,\n",
    "                    #whether P(birth) should be weighted by parental dist\n",
    "                    'dist_weighted_birth':       False,\n",
    "                    #intrinsic growth rate\n",
    "                    'R':                        0.5,\n",
    "                    #intrinsic birth rate (MUST BE 0<=b<=1)\n",
    "                    'b':                        0.2,\n",
    "                    #expectation of distr of n offspring per mating pair\n",
    "                    'n_births_distr_lambda':    1,\n",
    "                    #whether n births should be fixed at n_births_dist_lambda\n",
    "                    'n_births_fixed':           True,\n",
    "                    #ADDED BY AB: choose nearest mate\n",
    "                    'choose_nearest_mate':      False,\n",
    "                    #ADDED BY AB: choose nearest mate\n",
    "                    'inverse_dist_mating':      False,\n",
    "                    \n",
    "########################################################\n",
    "# CONSIDER CHANGING THIS PARAMETER                     #\n",
    "                    #radius of mate-search area        #   \n",
    "                    'mating_radius':            1,    #\n",
    "########################################################            \n",
    "                    \n",
    "                    }, # <END> 'mating'\n",
    "\n",
    "            #----------------------------------------#\n",
    "            #--- spp num. 0: mortality parameters ---#\n",
    "            #----------------------------------------#\n",
    "\n",
    "                'mortality'     : {\n",
    "                    #maximum age\n",
    "                    'max_age':                      None,\n",
    "                    #min P(death) (MUST BE 0<=d_min<=1)\n",
    "                    'd_min':                        0,\n",
    "                    #max P(death) (MUST BE 0<=d_max<=1)\n",
    "                    'd_max':                        1,\n",
    "                    #width of window used to estimate local pop density\n",
    "                    'density_grid_window_width':    None,\n",
    "                    }, # <END> 'mortality'\n",
    "\n",
    "            #---------------------------------------#\n",
    "            #--- spp num. 0: movement parameters ---#\n",
    "            #---------------------------------------#\n",
    "\n",
    "                'movement': {\n",
    "                    #whether or not the species is mobile\n",
    "                    'move':                                 True,\n",
    "                    #mode of distr of movement direction\n",
    "                    'direction_distr_mu':                   0,\n",
    "                    #concentration of distr of movement direction\n",
    "                    'direction_distr_kappa':                0,\n",
    "##################################################################\n",
    "# CONSIDER CHANGING THIS PARAMETER                               #\n",
    "                    #1st param of distr of movement distance  (mean)#   \n",
    "                    'movement_distance_distr_param1':       0.5, #\n",
    "##################################################################\n",
    "                    #2nd param of distr of movement distance (variance)\n",
    "                    'movement_distance_distr_param2':       5e-8,\n",
    "                    #movement distance distr to use ('levy' or 'wald')\n",
    "                    'movement_distance_distr':              'levy',\n",
    "                    #1st param of distr of dispersal distance\n",
    "                    'dispersal_distance_distr_param1':      0.5,\n",
    "                    #2nd param of distr of dispersal distance\n",
    "                    'dispersal_distance_distr_param2':      5e-14,\n",
    "                    #dispersal distance distr to use ('levy' or 'wald')\n",
    "                    'dispersal_distance_distr':             'levy',\n",
    "                    'move_surf'     : {\n",
    "                        #move-surf Layer name\n",
    "                        'layer':                'lyr_0',\n",
    "                        #whether to use mixture distrs\n",
    "                        'mixture':              True,\n",
    "                        #concentration of distrs\n",
    "                        'vm_distr_kappa':       12,\n",
    "                        #length of approximation vectors for distrs\n",
    "                        'approx_len':           5000,\n",
    "                        }, # <END> 'move_surf'\n",
    "\n",
    "                    },    # <END> 'movement'\n",
    "\n",
    "\n",
    "            #---------------------------------------------------#\n",
    "            #--- spp num. 0: genomic architecture parameters ---#\n",
    "            #---------------------------------------------------#\n",
    "\n",
    "                'gen_arch': {\n",
    "                    #file defining custom genomic arch\n",
    "                    'gen_arch_file':            None,\n",
    "                    #num of loci\n",
    "                    'L':                        1000,\n",
    "                    #starting allele frequency (None to draw freqs randomly)\n",
    "                    'start_p_fixed':            0.5,\n",
    "                    #whether to start neutral locus freqs at 0\n",
    "                    'start_neut_zero':          False,\n",
    "                    #genome-wide per-base neutral mut rate (0 to disable)\n",
    "                    'mu_neut':                  0,\n",
    "                    #genome-wide per-base deleterious mut rate (0 to disable)\n",
    "                    'mu_delet':                 0,\n",
    "                    #shape of distr of deleterious effect sizes\n",
    "                    'delet_alpha_distr_shape':  0.2,\n",
    "                    #scale of distr of deleterious effect sizes\n",
    "                    'delet_alpha_distr_scale':  0.2,\n",
    "                    #alpha of distr of recomb rates\n",
    "                    'r_distr_alpha':            None,\n",
    "                    #beta of distr of recomb rates\n",
    "                    'r_distr_beta':             None,\n",
    "                    #whether loci should be dominant (for allele '1')\n",
    "                    'dom':                      False,\n",
    "                    #whether to allow pleiotropy\n",
    "                    'pleiotropy':               False,\n",
    "                    #custom fn for drawing recomb rates\n",
    "                    'recomb_rate_custom_fn':    None,\n",
    "                    #number of recomb paths to hold in memory\n",
    "                    'n_recomb_paths_mem':       int(1e4),\n",
    "                    #total number of recomb paths to simulate\n",
    "                    'n_recomb_paths_tot':       int(1e5),\n",
    "                    #num of crossing-over events (i.e. recombs) to simulate\n",
    "                    'n_recomb_sims':            10_000,\n",
    "                    #whether to generate recombination paths at each timestep\n",
    "                    'allow_ad_hoc_recomb':      False,\n",
    "                    #whether to save mutation logs\n",
    "                    'mut_log':                  False,\n",
    "\n",
    "                    'traits': {\n",
    "\n",
    "                        #--------------------------#\n",
    "                        #--- trait 0 parameters ---#\n",
    "                        #--------------------------#\n",
    "                        #trait name (TRAIT NAMES MUST BE UNIQUE!)\n",
    "                        'trait_1': {\n",
    "                            #trait-selection Layer name\n",
    "                            'layer':                'lyr_1',\n",
    "#################################################################\n",
    "# CONSIDER CHANGING THESE PARAMETERS                            #\n",
    "                            #polygenic selection coefficient    #\n",
    "                            'phi':                  0.5,        # \n",
    "                            #number of loci underlying trait    #\n",
    "                            'n_loci':               5,         # \n",
    "#################################################################\n",
    "                            #mutation rate at loci underlying trait\n",
    "                            'mu':                   0,\n",
    "                            #mean of distr of effect sizes\n",
    "                            'alpha_distr_mu' :      0.1,\n",
    "                            #variance of distr of effect size\n",
    "                            'alpha_distr_sigma':    0,\n",
    "                            #max allowed magnitude for an alpha value\n",
    "                            'max_alpha_mag':        None,\n",
    "                            #curvature of fitness function\n",
    "                            'gamma':                1,\n",
    "                            #whether the trait is universally advantageous\n",
    "                            'univ_adv':             False\n",
    "                            }, # <END> trait 0\n",
    "                        \n",
    "                        #--------------------------#\n",
    "                        #--- trait 1 parameters ---#\n",
    "                        #--------------------------#\n",
    "                        #trait name (TRAIT NAMES MUST BE UNIQUE!)\n",
    "                        'trait_2': {\n",
    "                            #trait-selection Layer name\n",
    "                            'layer':                'lyr_2',\n",
    "#################################################################\n",
    "# CONSIDER CHANGING THESE PARAMETERS                            #\n",
    "                            #polygenic selection coefficient    #\n",
    "                            'phi':                  0.5,        # \n",
    "                            #number of loci underlying trait    #\n",
    "                            'n_loci':               5,         # \n",
    "#################################################################\n",
    "                            #mutation rate at loci underlying trait\n",
    "                            'mu':                   0,\n",
    "                            #mean of distr of effect sizes\n",
    "                            'alpha_distr_mu' :      0.1,\n",
    "                            #variance of distr of effect size\n",
    "                            'alpha_distr_sigma':    0,\n",
    "                            #max allowed magnitude for an alpha value\n",
    "                            'max_alpha_mag':        None,\n",
    "                            #curvature of fitness function\n",
    "                            'gamma':                1,\n",
    "                            #whether the trait is universally advantageous\n",
    "                            'univ_adv':             False\n",
    "                            }, # <END> trait 0\n",
    "\n",
    "\n",
    "    #### NOTE: Individual Traits' sections can be copy-and-pasted (and\n",
    "    #### assigned distinct keys and names), to create additional Traits.\n",
    "\n",
    "\n",
    "                        }, # <END> 'traits'\n",
    "\n",
    "                    }, # <END> 'gen_arch'\n",
    "\n",
    "\n",
    "                }, # <END> spp num. 0\n",
    "\n",
    "\n",
    "\n",
    "    #### NOTE: individual Species' sections can be copy-and-pasted (and\n",
    "    #### assigned distinct keys and names), to create additional Species.\n",
    "\n",
    "\n",
    "            }, # <END> 'species'\n",
    "\n",
    "        }, # <END> 'comm'\n",
    "\n",
    "\n",
    "#------------------------------------------------------------------------#\n",
    "\n",
    "#-------------#\n",
    "#--- MODEL ---#\n",
    "#-------------#\n",
    "    'model': {\n",
    "        #total Model runtime (in timesteps)\n",
    "        'T':            100,\n",
    "        #min burn-in runtime (in timesteps)\n",
    "        'burn_T':       30,\n",
    "        #seed number\n",
    "        'num':          None,\n",
    "        #time step interval for simplication of tskit tables\n",
    "        'tskit_simp_interval':      100,\n",
    "\n",
    "\n",
    "        #-----------------------------#\n",
    "        #--- iterations parameters ---#\n",
    "        #-----------------------------#\n",
    "        'its': {\n",
    "            #num iterations\n",
    "            'n_its':            1,\n",
    "            #whether to randomize Landscape each iteration\n",
    "            'rand_landscape':   False,\n",
    "            #whether to randomize Community each iteration\n",
    "            'rand_comm':        False,\n",
    "            #whether to burn in each iteration\n",
    "            'repeat_burn':      False,\n",
    "            }, # <END> 'iterations'\n",
    "        \n",
    "        ####################################\n",
    "        #### data-collection parameters ####\n",
    "        ####################################\n",
    "        'data': {\n",
    "            'sampling': {\n",
    "                #sampling scheme {'all', 'random', 'point', 'transect'}\n",
    "                'scheme':               'all',\n",
    "                #sample size at each point, for point & transect sampling\n",
    "                'n':                    None,\n",
    "                #coords of collection points, for point sampling\n",
    "                'points':               None,\n",
    "                #coords of transect endpoints, for transect sampling\n",
    "                'transect_endpoints':   None,\n",
    "                #num points along transect, for transect sampling\n",
    "                'n_transect_points':    None,\n",
    "                #collection radius around points, for point & transect sampling\n",
    "                'radius':               None,\n",
    "                #when to collect data\n",
    "                'when':                 0,\n",
    "                #whether to save current Layers when data is collected\n",
    "                'include_landscape':    False,\n",
    "                #whether to include fixed loci in VCF files\n",
    "                'include_fixed_sites':  False,\n",
    "                },\n",
    "            'format': {\n",
    "                #format for genetic data {'vcf', 'fasta'}\n",
    "                'gen_format':           ['vcf', 'fasta'],\n",
    "                #format for vector geodata {'csv', 'shapefile', 'geojson'}\n",
    "                'geo_vect_format':      'csv',\n",
    "                #format for raster geodata {'geotiff', 'txt'}\n",
    "                'geo_rast_format':      'geotiff',\n",
    "                },\n",
    "            }, #<END> 'data'\n",
    "        \n",
    "        #-----------------------------------#\n",
    "        #--- stats-collection parameters ---#\n",
    "        #-----------------------------------#\n",
    "        'stats': {\n",
    "            #number of individs at time t\n",
    "            'Nt': {\n",
    "                #whether to calculate\n",
    "                'calc':     True,\n",
    "                #calculation frequency (in timesteps)\n",
    "                'freq':     1,\n",
    "                },\n",
    "            #heterozgosity\n",
    "            'het': {\n",
    "                #whether to calculate\n",
    "                'calc':     True,\n",
    "                #calculation frequency (in timesteps)\n",
    "                'freq':     5,\n",
    "                #whether to mean across sampled individs\n",
    "                'mean': False,\n",
    "                },\n",
    "            #minor allele freq\n",
    "            'maf': {\n",
    "                #whether to calculate\n",
    "                'calc':     True,\n",
    "                #calculation frequency (in timesteps)\n",
    "                'freq':     5,\n",
    "                },\n",
    "            #mean fitness\n",
    "            'mean_fit': {\n",
    "                #whether to calculate\n",
    "                'calc':     True,\n",
    "                #calculation frequency (in timesteps)\n",
    "                'freq':     5,\n",
    "                },\n",
    "            #linkage disequilibirum\n",
    "            'ld': {\n",
    "                #whether to calculate\n",
    "                'calc':     False,\n",
    "                #calculation frequency (in timesteps)\n",
    "                'freq':     100,\n",
    "                },\n",
    "            }, # <END> 'stats'\n",
    "\n",
    "        } # <END> 'model'\n",
    "\n",
    "    } # <END> params\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.random.seed(42)\n",
    "#set K factor (pop size)\n",
    "params['comm']['species']['spp_0']['init']['K_factor'] = 5\n",
    "\n",
    "#fix mating radius\n",
    "params['comm']['species']['spp_0']['mating']['mating_radius'] = 1\n",
    "\n",
    "#Phi\n",
    "params['comm']['species']['spp_0']['gen_arch']['traits']['trait_0']['phi'] = 0.05\n",
    "\n",
    "#set md and dd\n",
    "params['comm']['species']['spp_0']['movement']['dispersal_distance_distr_param1'] = 1\n",
    "params['comm']['species']['spp_0']['movement']['movement_distance_distr_param1'] = 1\n",
    "\n",
    "#make our params dict into a proper Geonomics ParamsDict object\n",
    "params = gnx.make_params_dict(params, 'MNLM_K5')\n",
    "    \n",
    "#then use it to make a model\n",
    "mod = gnx.make_model(parameters=params, verbose=True)\n",
    "\n",
    "mod.walk(T=10000, mode='burn')\n",
    "mod.walk(100)\n",
    "\n",
    "plt.show()\n",
    "mod.plot_phenotype(spp = 0, trt = 0)\n",
    "gea_df = DataFrameGEA(mod, spp = 0)\n",
    "gea_df.to_csv(\"data/gea_MNLM_dd_md_1_phi0p05_k5_TEST_df.csv\")\n",
    "\n",
    "mod = mod_dict[str(m)]\n",
    "loci_df = pd.DataFrame(range(1,10001))\n",
    "\n",
    "trait1 = mod.comm[0].gen_arch.traits[0].loci\n",
    "loci_df['trait1'] = 0\n",
    "loci_df['trait1'][trait1] = 1\n",
    "\n",
    "trait2 = mod.comm[0].gen_arch.traits[1].loci\n",
    "loci_df['trait2'] = 0\n",
    "loci_df['trait2'][trait2] = 1\n",
    "\n",
    "loci_df.to_csv(\"data/loci_MNLM_m1_phi0p05_k5_H05_r06_TEST_df.csv\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
