{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import geonomics as gnx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "MAKING MODEL...\n",
      "\n",
      "\tMAKING LANDSCAPE...\n",
      "\n",
      "\tMAKING COMMUNITY...\n",
      "\n",
      "\t\tMAKING SPECIES Sceloporus graciosus...\n",
      "\n",
      "\t\t\tmaking genomic architecture...\n",
      "\n",
      "\t\t\tmaking individuals...\n",
      "\n",
      "\t\t\tmaking movement surface...\n",
      "\t\t\t\t[can take a bit]\n",
      "\n",
      "\t\t\tsetting up species changes...\n",
      "\t\t\t\t[can take a while,\n",
      "\t\t\t\t if movement or dispersal\n",
      "\t\t\t\t surfaces will change]\n",
      "\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-6-24e616726856>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mmod\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgnx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_demo\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'yosemite'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtime_it\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msave_figs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmake_gifs\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmake_3d_plots\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mn_individs_plot\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m20_000\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/.local/lib/python3.7/site-packages/geonomics/main.py\u001b[0m in \u001b[0;36mrun_demo\u001b[0;34m(name, save_figs, time_it, **kwargs)\u001b[0m\n\u001b[1;32m    722\u001b[0m         \u001b[0mparams\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdemos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_yosemite\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_params\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    723\u001b[0m         \u001b[0mparams\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmake_params_dict\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparams\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 724\u001b[0;31m         \u001b[0mmod\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mdemos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_yosemite\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_run\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparams\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msave_figs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtime_it\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    725\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    726\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.7/site-packages/geonomics/demos/_yosemite.py\u001b[0m in \u001b[0;36m_run\u001b[0;34m(params, save_figs, time_it, make_gifs, make_3d_plots, n_individs_plot)\u001b[0m\n\u001b[1;32m    795\u001b[0m                     \u001b[0;34m'gen_format'\u001b[0m\u001b[0;34m:\u001b[0m           \u001b[0;34m[\u001b[0m\u001b[0;34m'vcf'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'fasta'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    796\u001b[0m                     \u001b[0;31m#format for vector geodata {'csv', 'shapefile', 'geojson'}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 797\u001b[0;31m                     \u001b[0;34m'geo_vect_format'\u001b[0m\u001b[0;34m:\u001b[0m      \u001b[0;34m'csv'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    798\u001b[0m                     \u001b[0;31m#format for raster geodata {'geotiff', 'txt'}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    799\u001b[0m                     \u001b[0;34m'geo_rast_format'\u001b[0m\u001b[0;34m:\u001b[0m      \u001b[0;34m'geotiff'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.7/site-packages/geonomics/main.py\u001b[0m in \u001b[0;36mmake_model\u001b[0;34m(parameters, verbose)\u001b[0m\n\u001b[1;32m    603\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    604\u001b[0m         \u001b[0mname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparameters\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmodel\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 605\u001b[0;31m         \u001b[0mmod\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mModel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mparameters\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverbose\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mverbose\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    606\u001b[0m         \u001b[0;32mreturn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmod\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    607\u001b[0m     \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.7/site-packages/geonomics/sim/model.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, name, params, verbose)\u001b[0m\n\u001b[1;32m    118\u001b[0m         \u001b[0;31m#make the land and community objects\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    119\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mland\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_landscape\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_verbose\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 120\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcomm\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_make_community\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_verbose\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    121\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    122\u001b[0m         \u001b[0;31m#and make the self._never_been_run attribute False,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.7/site-packages/geonomics/sim/model.py\u001b[0m in \u001b[0;36m_make_community\u001b[0;34m(self, verbose)\u001b[0m\n\u001b[1;32m    426\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_make_community\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverbose\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    427\u001b[0m         comm = _make_community(self.land, self.params, burn=True,\n\u001b[0;32m--> 428\u001b[0;31m                                verbose=verbose)\n\u001b[0m\u001b[1;32m    429\u001b[0m         \u001b[0;32mreturn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcomm\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    430\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.7/site-packages/geonomics/structs/community.py\u001b[0m in \u001b[0;36m_make_community\u001b[0;34m(land, params, burn, verbose)\u001b[0m\n\u001b[1;32m    136\u001b[0m                              \u001b[0mspp_params\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mparams\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcomm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mspecies\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    137\u001b[0m                              burn=burn, verbose=verbose) for n, name in\n\u001b[0;32m--> 138\u001b[0;31m             enumerate(params.comm.species.keys())}\n\u001b[0m\u001b[1;32m    139\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mCommunity\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mland\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mspps\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    140\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.7/site-packages/geonomics/structs/community.py\u001b[0m in \u001b[0;36m<dictcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    135\u001b[0m     spps = {n: _make_species(land=land, name=name, idx=n,\n\u001b[1;32m    136\u001b[0m                              \u001b[0mspp_params\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mparams\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcomm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mspecies\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 137\u001b[0;31m                              burn=burn, verbose=verbose) for n, name in\n\u001b[0m\u001b[1;32m    138\u001b[0m             enumerate(params.comm.species.keys())}\n\u001b[1;32m    139\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mCommunity\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mland\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mspps\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.7/site-packages/geonomics/structs/species.py\u001b[0m in \u001b[0;36m_make_species\u001b[0;34m(land, name, idx, spp_params, burn, verbose)\u001b[0m\n\u001b[1;32m   2536\u001b[0m             \u001b[0mch_params\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2537\u001b[0m         \u001b[0;31m#make _SpeciesChanger and set it to the spp's changer attribute\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2538\u001b[0;31m         \u001b[0mspp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_changer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_SpeciesChanger\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mch_params\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mland\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mland\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2539\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2540\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mspp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.7/site-packages/geonomics/ops/change.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, spp, spp_change_params, land)\u001b[0m\n\u001b[1;32m    164\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    165\u001b[0m     \u001b[0;31m#call self._set_changes() to set self.changes and self.next_change\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 166\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_set_changes\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mland\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    167\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    168\u001b[0m     \u001b[0;31m#method to set the base_K attribute to spp.K\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.7/site-packages/geonomics/ops/change.py\u001b[0m in \u001b[0;36m_set_changes\u001b[0;34m(self, spp, land)\u001b[0m\n\u001b[1;32m    197\u001b[0m                 \u001b[0mstart_lyr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mland\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0msurf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlyr_num\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmixture\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msurf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmix\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    198\u001b[0m                 \u001b[0mkappa\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msurf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkappa\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mapprox_len\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msurf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapprox_len\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 199\u001b[0;31m                 change_params_one_lyr = lc_move_surf_params)\n\u001b[0m\u001b[1;32m    200\u001b[0m             \u001b[0;31m#and create change fns from them\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    201\u001b[0m             move_surf_change_fns.extend(\n",
      "\u001b[0;32m~/.local/lib/python3.7/site-packages/geonomics/ops/change.py\u001b[0m in \u001b[0;36m_make_conductance_surface_series\u001b[0;34m(land, start_lyr, mixture, kappa, approx_len, change_params_one_lyr)\u001b[0m\n\u001b[1;32m    591\u001b[0m         \u001b[0;31m#then create a Movement_Surface using the copied Landscape\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    592\u001b[0m         surf_series.append((t, spt._ConductanceSurface(dummy_lyr, mixture,\n\u001b[0;32m--> 593\u001b[0;31m             approx_len = approx_len, vm_distr_kappa = kappa)))\n\u001b[0m\u001b[1;32m    594\u001b[0m     \u001b[0;32mreturn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msurf_series\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    595\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.7/site-packages/geonomics/utils/spatial.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, cond_lyr, mixture, approx_len, vm_distr_kappa)\u001b[0m\n\u001b[1;32m    182\u001b[0m                                               \u001b[0mmixture\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmix\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    183\u001b[0m                                               \u001b[0mvm_distr_kappa\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkappa\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 184\u001b[0;31m                                               approx_len=self.approx_len)\n\u001b[0m\u001b[1;32m    185\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    186\u001b[0m         assert self.approx_len == self.surf.shape[2], (\"_ConductanceSurface.\"\n",
      "\u001b[0;32m~/.local/lib/python3.7/site-packages/geonomics/utils/spatial.py\u001b[0m in \u001b[0;36m_make_conductance_surface\u001b[0;34m(rast, mixture, approx_len, vm_distr_kappa)\u001b[0m\n\u001b[1;32m    463\u001b[0m                 cond_surf[i, j, :] = _make_von_mises_multimodal_sampler(\n\u001b[1;32m    464\u001b[0m                     \u001b[0mneigh\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mqueen_dirs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvm_distr_kappa\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mvm_distr_kappa\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 465\u001b[0;31m                     approx_len=approx_len)\n\u001b[0m\u001b[1;32m    466\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    467\u001b[0m                 cond_surf[i, j, :] = _make_von_mises_unimodal_sampler(\n",
      "\u001b[0;32m~/.local/lib/python3.7/site-packages/geonomics/utils/spatial.py\u001b[0m in \u001b[0;36m_make_von_mises_multimodal_sampler\u001b[0;34m(neigh, dirs, vm_distr_kappa, approx_len)\u001b[0m\n\u001b[1;32m    430\u001b[0m     approx = np.hstack([s_vonmises.rvs(vm_distr_kappa, loc=loc, scale=1,\n\u001b[1;32m    431\u001b[0m                                        \u001b[0msize\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msize\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 432\u001b[0;31m                                        ) for loc, size in loc_choices])\n\u001b[0m\u001b[1;32m    433\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mapprox\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    434\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/.local/lib/python3.7/site-packages/geonomics/utils/spatial.py\u001b[0m in \u001b[0;36m<listcomp>\u001b[0;34m(.0)\u001b[0m\n\u001b[1;32m    430\u001b[0m     approx = np.hstack([s_vonmises.rvs(vm_distr_kappa, loc=loc, scale=1,\n\u001b[1;32m    431\u001b[0m                                        \u001b[0msize\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msize\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 432\u001b[0;31m                                        ) for loc, size in loc_choices])\n\u001b[0m\u001b[1;32m    433\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mapprox\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    434\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/global/software/sl-7.x86_64/modules/langs/python/3.7/lib/python3.7/site-packages/scipy/stats/_distn_infrastructure.py\u001b[0m in \u001b[0;36mrvs\u001b[0;34m(self, *args, **kwds)\u001b[0m\n\u001b[1;32m    978\u001b[0m         \u001b[0;31m# by _rvs().\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    979\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_size\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msize\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 980\u001b[0;31m         \u001b[0mvals\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_rvs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    981\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    982\u001b[0m         \u001b[0mvals\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvals\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0mscale\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mloc\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/global/software/sl-7.x86_64/modules/langs/python/3.7/lib/python3.7/site-packages/scipy/stats/_continuous_distns.py\u001b[0m in \u001b[0;36m_rvs\u001b[0;34m(self, kappa)\u001b[0m\n\u001b[1;32m   7491\u001b[0m     \"\"\"\n\u001b[1;32m   7492\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_rvs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkappa\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 7493\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_random_state\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvonmises\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0.0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkappa\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msize\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_size\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   7494\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   7495\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_pdf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkappa\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "mod = gnx.run_demo('yosemite', time_it=True, save_figs=True, make_gifs=True, make_3d_plots=True, n_individs_plot=20_000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name '__file__' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-38e0d8ea97c7>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     33\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     34\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 35\u001b[0;31m \u001b[0mgnx_dir\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m__file__\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     36\u001b[0m \u001b[0mgnx_dir\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgnx_dir\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     37\u001b[0m \u001b[0mDATA_PATH\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpath\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgnx_dir\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"data\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"yosemite_demo\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name '__file__' is not defined"
     ]
    }
   ],
   "source": [
    "#!/usr/bin/python\n",
    "\n",
    "#####################\n",
    "#TODO\n",
    "\n",
    "#Add a \"timing\" argument, that I can set to false to run the model without\n",
    "#plotting, to get an accurate assessment of run time\n",
    "\n",
    "#####################\n",
    "\n",
    "# geonomics imports\n",
    "from geonomics.utils.viz import _check_display\n",
    "\n",
    "# other imports\n",
    "import os\n",
    "import numpy as np\n",
    "import matplotlib as mpl\n",
    "_check_display()\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import animation\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "from mpl_toolkits.mplot3d.axes3d import Axes3D\n",
    "import time\n",
    "import rasterio as rio\n",
    "\n",
    "try:\n",
    "    from pykrige.ok import OrdinaryKriging\n",
    "    with_pykrige = True\n",
    "except ModuleNotFoundError:\n",
    "    print((\"NOTE: Module 'pykrige' not found. Yosemite demo 3d plots cannot be\"\n",
    "          \" produced.\"))\n",
    "    with_pykrige = False\n",
    "\n",
    "\n",
    "gnx_dir = os.path.split(__file__)[0]\n",
    "gnx_dir = os.path.join(*os.path.split(gnx_dir)[:-1])\n",
    "DATA_PATH = os.path.join(gnx_dir, \"data\", \"yosemite_demo\")\n",
    "\n",
    "\n",
    "def krig_phenotype(mod, window_width=8, rel_bandwidth=0.5, max_n_pts=10000):\n",
    "    print('Kriging phenotype. Please wait...\\n\\n')\n",
    "    if len(mod.comm[0]) > max_n_pts:\n",
    "        inds = np.random.choice([*mod.comm[0].keys()], size=5000,\n",
    "                                replace=False)\n",
    "    else:\n",
    "        inds = np.array([*mod.comm[0].keys()])\n",
    "    xs = mod.comm[0]._get_x(inds)\n",
    "    ys = mod.comm[0]._get_y(inds)\n",
    "    zs = mod.comm[0]._get_z(individs=inds)[:,0]\n",
    "    OK = OrdinaryKriging(xs, ys, zs, variogram_model='spherical')\n",
    "    gridx = np.arange(0.5, mod.land.dim[0] + 0.5, 1)\n",
    "    gridy = np.arange(0.5, mod.land.dim[1] + 0.5, 1)\n",
    "    krig_pheno, ss = OK.execute('grid', gridx, gridy)\n",
    "    return krig_pheno\n",
    "\n",
    "\n",
    "# plot 3 change rasters, 1 for each of the 3 landscape layers\n",
    "def make_change_fig(mod):\n",
    "    change_fig = plt.figure()\n",
    "    change_fig_titles = {0: 'temperature',\n",
    "                         2: 'precipitation',\n",
    "                         1: 'habitat suitability',\n",
    "                     }\n",
    "    cmaps = ['afmhot_r', 'rainbow_r', 'PiYG']\n",
    "    for i, d in enumerate(change_fig_titles.items()):\n",
    "        ax = change_fig.add_subplot(1, 3, i + 1)\n",
    "        lyr_num = d[0]\n",
    "        title = d[1]\n",
    "        ax.set_title(title, size=20)\n",
    "        af = mod.land[lyr_num]._get_rast_in_native_units()\n",
    "        b4 = mod.orig_land[lyr_num]._get_rast_in_native_units()\n",
    "        diff = af - b4\n",
    "        im = ax.imshow(diff, cmap = cmaps[i])\n",
    "        units = '$\\Delta$ ' + mod.land[lyr_num].units\n",
    "        cb = change_fig.colorbar(im, orientation='horizontal')\n",
    "        cb.set_label(label=units, size=13)\n",
    "        cb.ax.tick_params(labelsize=9)\n",
    "        (xtick_locs, xtick_labs, ytick_locs,\n",
    "         ytick_labs) = mod.land[lyr_num]._get_coord_ticks()\n",
    "        ax.set_xticks(xtick_locs)\n",
    "        ax.set_xticklabels(np.round(xtick_labs, 1), rotation=45, size=7,\n",
    "                           color='gray')\n",
    "        ax.set_yticks(ytick_locs)\n",
    "        ax.set_yticklabels(ytick_labs, size=7, color='gray')\n",
    "        ax.set_xlabel('lon', color='gray', size=10)\n",
    "        if i == 0:\n",
    "            ax.set_ylabel('lat', color='gray', size=10)\n",
    "    change_fig.show()\n",
    "    return change_fig\n",
    "\n",
    "\n",
    "# drape a raster on top of a DEM\n",
    "def drape_raster(mod, rast, DEM, cbar_label, gif_filename,\n",
    "                 save_figs, make_gifs, cmap='rainbow'):\n",
    "    fig3d = plt.figure()\n",
    "    ax3d = fig3d.add_subplot(111, projection='3d')\n",
    "    ax3d.set_xlabel('lon', color='gray', size=10)\n",
    "    ax3d.set_ylabel('lat', color='gray', size=10)\n",
    "    ax3d.set_zlabel('alt (m)', color='gray', size=10)\n",
    "    # get the x and y coordinates for each cell\n",
    "    xi = range(mod.land.dim[0])\n",
    "    yi = range(mod.land.dim[1])\n",
    "    # use the x and y coords to get a meshgrid\n",
    "    X, Y = np.meshgrid(xi, yi)\n",
    "    # create a mappable object, to use to make the colorbar\n",
    "    mappable = plt.cm.ScalarMappable()\n",
    "    mappable.set_array(rast)\n",
    "    mappable.set_cmap(cmap)\n",
    "\n",
    "    # initialization function, which plots the background of each frame\n",
    "    def init():\n",
    "        # plot the DEM as a surface, coloring its patches by the draped raster\n",
    "        surf = ax3d.plot_surface(X, Y, DEM, facecolors=mappable.cmap(rast),\n",
    "                                 cmap=mappable.cmap)\n",
    "        # add a 2d contour plot projected onto the horizontal below it\n",
    "        cont = ax.contourf(X, Y, rast, zdir='elevation', offset=np.min(DEM),\n",
    "                           cmap=mappable.cmap)\n",
    "        # create a colorbar\n",
    "        cb = plt.colorbar(mappable)\n",
    "        cb.set_label(label=cbar_label, size=13)\n",
    "        cb.ax.tick_params(labelsize=9)\n",
    "        # set the x-, y-, and z-axis ticks and labels\n",
    "        (xtlocs, xtlabs, ytlocs, ytlabs) = mod.land[0]._get_coord_ticks()\n",
    "        ztlocs = ax.zaxis.get_ticklocs()\n",
    "        ztlabs = [int(txt.get_text()) for txt in ax.zaxis.get_ticklabels()]\n",
    "        ax.set_xticks(xtlocs)\n",
    "        ax.set_xticklabels(np.round(xtlabs, 1), size=7, color='gray')\n",
    "        ax.set_yticks(ytlocs)\n",
    "        ax.set_yticklabels(np.round(ytlabs, 1), size=7, color='gray')\n",
    "        ax.set_zticks(ztlocs)\n",
    "        ax.set_zticklabels(np.round(ztlabs, 0), size=7, color='gray')\n",
    "\n",
    "        return fig3d,\n",
    "\n",
    "    if make_gifs:\n",
    "        # animation function, which will be called sequentially\n",
    "        def animate(i):\n",
    "            ax3d.view_init(elev=5., azim=i)\n",
    "            return fig3d,\n",
    "        anim = animation.FuncAnimation(fig3d, animate, init_func=init, frames=359,\n",
    "                                       interval=5, blit=True)\n",
    "    else:\n",
    "        ax3d.view_init(elev=5., azim=115)\n",
    "        init()\n",
    "\n",
    "    if save_figs:\n",
    "        try:\n",
    "            anim.save(gif_filename, writer='imagemagick', fps=60)\n",
    "        except Exception as e:\n",
    "            print(('\\nCould not use Imagemagick to save the 3D plot\\'s '\n",
    "                   'GIF.  The following error was thrown:\\n\\t%s') % e)\n",
    "\n",
    "    fig3d.show()\n",
    "\n",
    "    return fig3d\n",
    "\n",
    "\n",
    "# create and save images that Imagemagick will stitch into a GIF\n",
    "def save_model_gif(mod, save_figs, make_gifs, n_individs=5000):\n",
    "    # set up second figure, for a GIF animation\n",
    "    mid_col_width = 0.8\n",
    "    scnd_col_width = 1\n",
    "    gif_fig = plt.figure(figsize=(6.75 + mid_col_width * 6.75, 5.4))\n",
    "    ax2_gs = gif_fig.add_gridspec(1, 3, width_ratios=[1, scnd_col_width, 1])\n",
    "    # gif_fig, axs2 = plt.subplots(1, 2)\n",
    "    # gif_fig.set_tight_layout(True)\n",
    "    ax2_1 = gif_fig.add_subplot(ax2_gs[0, 0])\n",
    "    ax2_2 = gif_fig.add_subplot(ax2_gs[0, 2])\n",
    "    gif_fig.suptitle('Time step: %i' % (mod.t+1), size=20)\n",
    "    ax2_1.set_title('Temperature', size=12)\n",
    "    ax2_2.set_title('Habitat suitability', size=12)\n",
    "    ax2_1.set_xlabel('lon', size=8)\n",
    "    ax2_1.set_ylabel('lat', size=8)\n",
    "    ax2_2.set_xlabel('lon', size=8)\n",
    "    ax2_2.set_ylabel('lat', size=8)\n",
    "    ax2_1.set_aspect(1)\n",
    "    ax2_2.set_aspect(1)\n",
    "    # plot the population on both the temp and hab rasters\n",
    "    land1 = ax2_1.pcolormesh(mod.land._x_cell_bds,\n",
    "                             mod.land._y_cell_bds,\n",
    "                             mod.land[0]._get_rast_in_native_units(),\n",
    "                             cmap='coolwarm',\n",
    "                             vmin=mod.land[0]._scale_min,\n",
    "                             vmax=mod.land[0]._scale_max)\n",
    "    xticks, xticklabs, yticks, yticklabs = mod.land[0]._get_coord_ticks()\n",
    "    ax2_1.set_xticklabels(labels=xticklabs, size=6)\n",
    "    ax2_1.set_yticklabels(labels=yticklabs, size=6)\n",
    "    divider = make_axes_locatable(ax2_1)\n",
    "    cax = divider.append_axes(\"right\", size=\"5%\", pad=0.05)\n",
    "    cbar_ticks = mod.land[0]._get_cbar_ticks_and_minmax_scaled_vals()\n",
    "    cbar = plt.colorbar(land1, cax=cax, ticks=np.linspace(cbar_ticks[1],\n",
    "                                                          cbar_ticks[2], 5))\n",
    "    cbar.ax.set_yticklabels(cbar_ticks[0], size=6)\n",
    "    cbar.set_label(mod.land[0].units, rotation=270, labelpad=5, y=0.5,\n",
    "                   size=7)\n",
    "    # get a random subset of n_individs individuals, to use in the plot\n",
    "    individs = mod.comm[0]._get_random_individuals(n_individs)\n",
    "    coords = mod.comm[0]._get_plot_coords(individs=individs)\n",
    "    xs = coords[:, 0]\n",
    "    ys = coords[:, 1]\n",
    "    zs = mod.comm[0]._get_z(individs=individs)[:, 0]\n",
    "    ax2_1.scatter(xs, ys, c=zs, s=1, cmap='coolwarm')\n",
    "    land2 = ax2_2.pcolormesh(mod.land._x_cell_bds,\n",
    "                             mod.land._y_cell_bds,\n",
    "                             mod.land[1]._get_rast_in_native_units(),\n",
    "                             cmap='BrBG_r',\n",
    "                             vmin=mod.land[1]._scale_min,\n",
    "                             vmax=mod.land[1]._scale_max)\n",
    "    xticks, xticklabs, yticks, yticklabs = mod.land[1]._get_coord_ticks()\n",
    "    ax2_2.set_xticklabels(labels=xticklabs, size=6)\n",
    "    ax2_2.set_yticklabels(labels=yticklabs, size=6)\n",
    "    divider = make_axes_locatable(ax2_2)\n",
    "    cax = divider.append_axes(\"right\", size=\"5%\", pad=0.05)\n",
    "    cbar_ticks = mod.land[1]._get_cbar_ticks_and_minmax_scaled_vals()\n",
    "    cbar = plt.colorbar(land2, cax=cax, ticks=np.linspace(cbar_ticks[1],\n",
    "                                                          cbar_ticks[2], 5))\n",
    "    cbar.ax.set_yticklabels(cbar_ticks[0], size=6)\n",
    "    cbar.set_label(mod.land[1].units, rotation=270, labelpad=15, y=0.5,\n",
    "                   size=10)\n",
    "    ax2_2.scatter(xs, ys, c='black', s=0.5)\n",
    "\n",
    "    # save the figure\n",
    "    if save_figs and make_gifs:\n",
    "        gif_fig.savefig('gif_img_%i.jpg' % (mod.t + 1))\n",
    "\n",
    "    # manually close the figure\n",
    "    plt.close(gif_fig)\n",
    "\n",
    "\n",
    "\n",
    "def _make_params():\n",
    "\n",
    "    # This is a parameters file generated by Geonomics\n",
    "    # (by the gnx.make_parameters_file() function).\n",
    "\n",
    "\n",
    "                       ##  ::::::          :::    :: ::::::::::##\n",
    "                 ##:::::    ::::   :::      ::    :: :: ::::::::::: :##\n",
    "              ##::::::::     ::            ::   ::::::::::::::::::::::::##\n",
    "            ##:::::::::                      :::::::::: :::::: ::::::::  :##\n",
    "          ## : ::::  ::                    ::::  : ::    :::::::: : ::  :   ##\n",
    "         ##GGGGG  EEEEE OOOOO   NN   NN   OOOOO   MM   MM IIIIII  CCCCC SSSSS##\n",
    "        ##GG     EE    OO   OO  NNN  NN  OO   OO  MM   MM   II   CC     SS    ##\n",
    "        ##GG     EE   OO     OO NN N NN OO     OO MMM MMM   II   CC     SSSSSS##\n",
    "        ##GG GGG EEEE OO     OO NN  NNN OO     OO MM M MM   II   CC         SS##\n",
    "        ##GG   G EE    OO   OO  NN   NN  OO   OO  MM   MM   II   CC        SSS##\n",
    "         ##GGGGG  EEEEE OOOOO   NN   NN   OOOOO   MM   MM IIIIII  CCCCC SSSSS##\n",
    "          ##    :::::::::               :::::::::: ::              ::  :   :##\n",
    "            ##:   :::::                    :::::: :::             ::::::: ##\n",
    "              ##   :::                      :::::  ::              :::::##\n",
    "                 ## ::                      ::::                     ##\n",
    "                       ##                                      ##\n",
    "                          ## :: ::    :::            ##\n",
    "\n",
    "\n",
    "    params = {\n",
    "    ###############################################################################\n",
    "\n",
    "    ###################\n",
    "    #### LANDSCAPE ####\n",
    "    ###################\n",
    "        'landscape': {\n",
    "\n",
    "        ##############\n",
    "        #### main ####\n",
    "        ##############\n",
    "            'main': {\n",
    "\n",
    "                # NOTE: The resolution of the rasters is 0.00833 dec. deg.,\n",
    "                # which is equal to about 0.00833* 88070.4 = 730.984 m in E-W\n",
    "                # and about 0.00833 * 111320 = 927.2956 m in N-S directions.\n",
    "                # NOTE: Thus, a cell is about 730.984 * 927.2956 = 678000 m^2,\n",
    "                # which is = 67.8 hectares.\n",
    "                # NOTE: The difference in distance in E-W and N-S is problematic\n",
    "                # for distance-based things!!!!\n",
    "\n",
    "                #y,x (a.k.a. i,j) dimensions of the Landscape\n",
    "                'dim':                      (90, 90),\n",
    "                #resolution of the Landscape\n",
    "                'res':                      (0, 0),\n",
    "                #upper-left corner of the Landscape\n",
    "                'ulc':                      (0, 0),\n",
    "                #projection of the Landscape\n",
    "                'prj':                      None,\n",
    "                }, # <END> 'main'\n",
    "\n",
    "        ################\n",
    "        #### layers ####\n",
    "        ################\n",
    "            'layers': {\n",
    "\n",
    "                #layer name (LAYER NAMES MUST BE UNIQUE!)\n",
    "                'tmp': {\n",
    "\n",
    "            #######################################\n",
    "            #### layer num. 0: init parameters ####\n",
    "            #######################################\n",
    "\n",
    "                    #initiating parameters for this layer\n",
    "                    'init': {\n",
    "\n",
    "                        #parameters for a 'file'-type Layer\n",
    "                        'file': {\n",
    "                            #</path/to/file>.<ext>\n",
    "                            'filepath': os.path.join(DATA_PATH,\n",
    "                                                     'yosemite_lyrs',\n",
    "                                                     'tmp_1980-2010_90x90.tif'),\n",
    "                            #minimum value to use to rescale the Layer to [0,1]\n",
    "                            'scale_min_val':                -1.37,\n",
    "                            #maximum value to use to rescale the Layer to [0,1]\n",
    "                            'scale_max_val':                19.75773,\n",
    "                            #decimal precision to use for coord-units (ulc & res)\n",
    "                            'coord_prec':                   8,\n",
    "                            #units of this file's variable\n",
    "                            'units':                    '$^∘C$',\n",
    "\n",
    "                            }, # <END> 'file'\n",
    "\n",
    "                        }, # <END> 'init'\n",
    "\n",
    "                #########################################\n",
    "                #### layer num. 0: change parameters ####\n",
    "                #########################################\n",
    "\n",
    "                    #landscape-change events for this Layer\n",
    "                    'change': {\n",
    "\n",
    "                        0: {\n",
    "                            #array or file for final raster of event, or directory\n",
    "                            #of files for each stepwise change in event\n",
    "                            'change_rast': os.path.join(DATA_PATH,\n",
    "                                                        'yosemite_lyrs',\n",
    "                                                        'tmp'),\n",
    "\n",
    "                            #starting timestep of event\n",
    "                            'start_t':          509,\n",
    "                            #ending timestep of event\n",
    "                            'end_t':            594,\n",
    "                            #number of stepwise changes in event\n",
    "                            'n_steps':          18,\n",
    "                            }, # <END> event 0\n",
    "\n",
    "                        }, # <END> 'change'\n",
    "\n",
    "                    }, # <END> layer num. 0\n",
    "\n",
    "\n",
    "                #layer name (LAYER NAMES MUST BE UNIQUE!)\n",
    "                'hab': {\n",
    "\n",
    "            #######################################\n",
    "            #### layer num. 1: init parameters ####\n",
    "            #######################################\n",
    "\n",
    "                    #initiating parameters for this layer\n",
    "                    'init': {\n",
    "\n",
    "                        #parameters for a 'file'-type Layer\n",
    "                        'file': {\n",
    "                            #</path/to/file>.<ext>\n",
    "                            'filepath': os.path.join(DATA_PATH,\n",
    "                                                     'yosemite_lyrs',\n",
    "                                                     'sdm_1980-2010_90x90.tif'),\n",
    "                            #minimum value to use to rescale the Layer to [0,1]\n",
    "                            'scale_min_val':                0,\n",
    "                            #maximum value to use to rescale the Layer to [0,1]\n",
    "                            'scale_max_val':                1,\n",
    "                            #decimal precision to use for coord-units (ulc & res)\n",
    "                            'coord_prec':                   8,\n",
    "                            #units of this file's variable\n",
    "                            'units':                    'suitability',\n",
    "\n",
    "                            }, # <END> 'file'\n",
    "\n",
    "                        }, # <END> 'init'\n",
    "\n",
    "                #########################################\n",
    "                #### layer num. 1: change parameters ####\n",
    "                #########################################\n",
    "\n",
    "                    #landscape-change events for this Layer\n",
    "                    'change': {\n",
    "\n",
    "                        0: {\n",
    "                            #array or file for final raster of event, or directory\n",
    "                            #of files for each stepwise change in event\n",
    "                            'change_rast': os.path.join(DATA_PATH,\n",
    "                                                        'yosemite_lyrs',\n",
    "                                                        'sdm'),\n",
    "\n",
    "                            #starting timestep of event\n",
    "                            'start_t':          509,\n",
    "                            #ending timestep of event\n",
    "                            'end_t':            594,\n",
    "                            #number of stepwise changes in event\n",
    "                            'n_steps':          18,\n",
    "                            }, # <END> event 0\n",
    "\n",
    "                        }, # <END> 'change'\n",
    "\n",
    "                    }, # <END> layer num. 1\n",
    "\n",
    "\n",
    "                #layer name (LAYER NAMES MUST BE UNIQUE!)\n",
    "                'ppt': {\n",
    "\n",
    "            #######################################\n",
    "            #### layer num. 2: init parameters ####\n",
    "            #######################################\n",
    "\n",
    "                    #initiating parameters for this layer\n",
    "                    'init': {\n",
    "\n",
    "                        #parameters for a 'file'-type Layer\n",
    "                        'file': {\n",
    "                            #</path/to/file>.<ext>\n",
    "                            'filepath': os.path.join(DATA_PATH,\n",
    "                                                     'yosemite_lyrs',\n",
    "                                                     'ppt_1980-2010_90x90.tif'),\n",
    "                            #minimum value to use to rescale the Layer to [0,1]\n",
    "                            'scale_min_val':                81.53713,\n",
    "                            #maximum value to use to rescale the Layer to [0,1]\n",
    "                            'scale_max_val':                2171.341,\n",
    "                            #decimal precision to use for coord-units (ulc & res)\n",
    "                            'coord_prec':                   8,\n",
    "                            #units of this file's variable\n",
    "                            'units':                    '$mm/yr$',\n",
    "\n",
    "                            }, # <END> 'file'\n",
    "\n",
    "                        }, # <END> 'init'\n",
    "\n",
    "                #########################################\n",
    "                #### layer num. 0: change parameters ####\n",
    "                #########################################\n",
    "\n",
    "                    #landscape-change events for this Layer\n",
    "                    'change': {\n",
    "\n",
    "                        0: {\n",
    "                            #array or file for final raster of event, or directory\n",
    "                            #of files for each stepwise change in event\n",
    "                            'change_rast': os.path.join(DATA_PATH,\n",
    "                                                        'yosemite_lyrs',\n",
    "                                                        'ppt'),\n",
    "\n",
    "                            #starting timestep of event\n",
    "                            'start_t':          509,\n",
    "                            #ending timestep of event\n",
    "                            'end_t':            594,\n",
    "                            #number of stepwise changes in event\n",
    "                            'n_steps':          18,\n",
    "                            }, # <END> event 0\n",
    "\n",
    "                        }, # <END> 'change'\n",
    "\n",
    "                    }, # <END> layer num. 0\n",
    "\n",
    "\n",
    "\n",
    "        #### NOTE: Individual Layers' sections can be copy-and-pasted (and\n",
    "        #### assigned distinct keys and names), to create additional Layers.\n",
    "\n",
    "\n",
    "                } # <END> 'layers'\n",
    "\n",
    "            }, # <END> 'landscape'\n",
    "\n",
    "\n",
    "    ###############################################################################\n",
    "\n",
    "    ###################\n",
    "    #### COMMUNITY ####\n",
    "    ###################\n",
    "        'comm': {\n",
    "\n",
    "            'species': {\n",
    "\n",
    "                #species name (SPECIES NAMES MUST BE UNIQUE!)\n",
    "                'Sceloporus graciosus': {\n",
    "\n",
    "                #####################################\n",
    "                #### spp num. 0: init parameters ####\n",
    "                #####################################\n",
    "\n",
    "                    'init': {\n",
    "                        #starting number of individs\n",
    "                        'N':                50000,\n",
    "                        #carrying-capacity Layer name\n",
    "                        'K_layer':          'hab',\n",
    "                        #multiplicative factor for carrying-capacity layer\n",
    "                        # NOTE: each cell is ~53.43 hectares, and pop density\n",
    "                        # estimates in Utah (closest estimate I could find to\n",
    "                        # Yosemite region) were 208/hectare, with little variation\n",
    "                        # (Tinkle 1973).\n",
    "                        # That would suggest each cell should have a carrying\n",
    "                        # capacity of ~ 53.43 * 208. However, in the Sierra they \n",
    "                        # and S. occidentalis segregate by habitat (Rose 1976),\n",
    "                        # so if we assume that only about 10% of land area\n",
    "                        # is covered by S. graciosus' preferred open,\n",
    "                        # exposed habitat, then that suggests we should\n",
    "                        # use a K_factor of 67.8 * 208 * 0.1 = 1410.24\n",
    "                        # NOTE: dividing by 10 for tractability\n",
    "                        'K_factor':         1410.24 / 10,\n",
    "                        }, # <END> 'init'\n",
    "\n",
    "                #######################################\n",
    "                #### spp num. 0: mating parameters ####\n",
    "                #######################################\n",
    "\n",
    "                    'mating'    : {\n",
    "                        #age(s) at sexual maturity (if tuple, female first)\n",
    "                        # NOTE: average reproductive age is 2 (i.e. in the third\n",
    "                        # season) (Tinkle 1973; Tinkle et. al 1993)\n",
    "                        'repro_age':                2,\n",
    "                        #whether to assign sexes\n",
    "                        'sex':                      True,\n",
    "                        # NOTE: Some studies suggest a skewed sex ratio\n",
    "                        # becuase of uneven male/female survival rate ratio\n",
    "                        # (e.g. Tinkle 1973), but other, more recent studies do not\n",
    "                        # (e.g. Tinkle et al. 1993), so just leaving sex_ratio at 1\n",
    "                        #ratio of males to females\n",
    "                        'sex_ratio':                1/1,\n",
    "                        # NOTE: found no information about this, but 0.5 seems\n",
    "                        # a reasonable default value\n",
    "                        #intrinsic growth rate\n",
    "                        'R':                        0.5,\n",
    "                        #intrinsic birth rate (MUST BE 0<=b<=1)\n",
    "                        'b':                        1,\n",
    "                        #expectation of distr of n offspring per mating pair\n",
    "                        #NOTE: using a lambda of equal to\n",
    "                        # avg_clutch_size * avg_n_clutches_per_yr * avg_surv_rate_to_yr_1\n",
    "                        # which I put at 4.464 * 2 * 0.16 = 1.428\n",
    "                        # (where the first value is avg of all CA pops in Tinkle et al 1993,\n",
    "                        # second value is from Tinkle 1973 and Tinkel et al 1993, and\n",
    "                        # third comes from the Mt. Diablo population in Ruth 1978,\n",
    "                        #cited in Tinkle et al. 1993)\n",
    "                        'n_births_distr_lambda':    1.428,\n",
    "                        #whether n births should be fixed at n_births_dist_lambda\n",
    "                        'n_births_fixed':           False,\n",
    "                        #radius of mate-search area\n",
    "                        #NOTE: just have individuals mate close by (i.e. within\n",
    "                        #12.457m, the average interannual movement distance\n",
    "                        'mating_radius':            0.5,\n",
    "                        'choose_nearest_mate':      False,\n",
    "                        'inverse_dist_mating':      False,\n",
    "                        }, # <END> 'mating'\n",
    "\n",
    "                ##########################################\n",
    "                #### spp num. 0: mortality parameters ####\n",
    "                ##########################################\n",
    "\n",
    "                    'mortality'     : {\n",
    "                        #maximum age\n",
    "                        # NOTE: 8 is the max age mentioned in Stebbins 1948's\n",
    "                        # Lassen population study\n",
    "                        'max_age':                      8,\n",
    "                        #min P(death) (MUST BE 0<=d_min<=1)\n",
    "                        'd_min':                        0,\n",
    "                        #max P(death) (MUST BE 0<=d_max<=1)\n",
    "                        'd_max':                        1,\n",
    "                        #width of window used to estimate local pop density\n",
    "                        'density_grid_window_width':    None,\n",
    "                        }, # <END> 'mortality'\n",
    "\n",
    "                #########################################\n",
    "                #### spp num. 0: movement parameters ####\n",
    "                #########################################\n",
    "\n",
    "                    'movement': {\n",
    "                        #NOTE: the movement and dispersal Wald distributions\n",
    "                        # are parameterized based on the finding of\n",
    "                        # Stebbins 1948 that these lizards have extremely\n",
    "                        # small home ranges;\n",
    "                        # expressed in cell-widths (where the cell width is\n",
    "                        # ~730.984 m, according to raster res converted\n",
    "                        # from dec. deg to m; see NOTE at top of file);\n",
    "                        # so becuase the mean dispersal distance of\n",
    "                        # Stebbins' lizards in that study is 12.457m (where\n",
    "                        # \"within/at edge of home area was treated as 0),\n",
    "                        # this would be 0.01704 cell widths;\n",
    "                        # then I assigned a sigma that allows for occasional\n",
    "                        # longer-distance movement events,\n",
    "                        # but not for long-distance offspring dispersal events\n",
    "\n",
    "                        #whether or not the species is mobile\n",
    "                        'move':                     True,\n",
    "                        #mode of distr of movement direction\n",
    "                        'direction_distr_mu':       1,\n",
    "                        #concentration of distr of movement direction\n",
    "                        'direction_distr_kappa':    0,\n",
    "                        #mean of distr of movement distance\n",
    "                        'movement_distance_distr_param1':        7e-5,\n",
    "                        #variance of distr of movement distance\n",
    "                        'movement_distance_distr_param2':     0.3,\n",
    "                        #mean of distr of dispersal distance\n",
    "                        'movement_distance_distr':      'lognormal',\n",
    "                        'dispersal_distance_distr_param1':       7.5e-4,\n",
    "                        #variance of distr of dispersal distance\n",
    "                        'dispersal_distance_distr_param2':    1,\n",
    "                        'dispersal_distance_distr':         'lognormal',\n",
    "                        'move_surf'     : {\n",
    "                            #move-surf Layer name\n",
    "                            'layer':                'hab',\n",
    "                            #whether to use mixture distrs\n",
    "                            'mixture':              True,\n",
    "                            #concentration of distrs\n",
    "                            'vm_distr_kappa':       12,\n",
    "                            #length of approximation vectors for distrs\n",
    "                            'approx_len':           5000,\n",
    "                            }, # <END> 'move_surf'\n",
    "\n",
    "                        },    # <END> 'movement'\n",
    "\n",
    "\n",
    "                #####################################################\n",
    "                #### spp num. 0: genomic architecture parameters ####\n",
    "                #####################################################\n",
    "\n",
    "                    'gen_arch': {\n",
    "                        #file defining custom genomic arch\n",
    "                        'gen_arch_file':            None,\n",
    "                        #num of loci\n",
    "                        'L':                        1000,\n",
    "                        #starting allele freq (None to draw freqs randomly)\n",
    "                        #at 0.5\n",
    "                        'start_p_fixed':            0.5,\n",
    "                        #genome-wide per-base neutral mut rate (0 to disable)\n",
    "                        'mu_neut':                  0,\n",
    "                        #genome-wide per-base deleterious mut rate\n",
    "                        #(0 to disable)\n",
    "                        'mu_delet':                 0,\n",
    "                        #shape of distr of deleterious effect sizes\n",
    "                        'delet_alpha_distr_shape':  0.01,\n",
    "                        #scale of distr of deleterious effect sizes\n",
    "                        'delet_alpha_distr_scale':  100,\n",
    "                        #alpha of distr of recomb rates\n",
    "                        'r_distr_alpha':            None,\n",
    "                        #beta of distr of recomb rates\n",
    "                        'r_distr_beta':             None,\n",
    "                        #whether loci should be dominant (for allele '1')\n",
    "                        'dom':                      False,\n",
    "                        #whether to allow pleiotropy\n",
    "                        'pleiotropy':               False,\n",
    "                        #custom fn for drawing recomb rates\n",
    "                        'recomb_rate_custom_fn':    None,\n",
    "                        #number of recomb paths to hold in memory\n",
    "                        'n_recomb_paths_mem':       int(1e4),\n",
    "                        #total number of recomb paths to simulate\n",
    "                        'n_recomb_paths_tot':       int(1e5),\n",
    "                        'n_recomb_sims':            10000,\n",
    "                        'start_neut_zero':          False,\n",
    "                        'allow_ad_hoc_recomb':      False,\n",
    "                        'jitter_breakpoints':       False,\n",
    "                        #whether to save mutation logs\n",
    "                        'mut_log':                  False,\n",
    "\n",
    "                        'traits': {\n",
    "\n",
    "                            ###########################\n",
    "                            ####trait 0 parameters ####\n",
    "                            ###########################\n",
    "                            #trait name (TRAIT NAMES MUST BE UNIQUE!)\n",
    "                            'thermal_tol': {\n",
    "                                #trait-selection Layer name\n",
    "                                'layer':                'tmp',\n",
    "                                #polygenic selection coefficient\n",
    "                                'phi':                  0.5,\n",
    "                                #number of loci underlying trait\n",
    "                                'n_loci':               100,\n",
    "                                #mutation rate at loci underlying trait\n",
    "                                'mu':                   0,\n",
    "                                #mean of distr of effect sizes\n",
    "                                'alpha_distr_mu' :      0.1,\n",
    "                                #variance of distr of effect size\n",
    "                                'alpha_distr_sigma':    0,\n",
    "                                #max allowed magnitude for an alpha value\n",
    "                                'max_alpha_mag':        None,\n",
    "                                #curvature of fitness function\n",
    "                                'gamma':                1,\n",
    "                                #whether the trait is universally advantageous\n",
    "                                'univ_adv':             False\n",
    "                                }, # <END> trait 0\n",
    "\n",
    "\n",
    "        #### NOTE: Individual Traits' sections can be copy-and-pasted (and\n",
    "        #### assigned distinct keys and names), to create additional Traits.\n",
    "\n",
    "\n",
    "                            }, # <END> 'traits'\n",
    "\n",
    "                        }, # <END> 'gen_arch'\n",
    "\n",
    "\n",
    "                    }, # <END> spp num. 0\n",
    "\n",
    "\n",
    "\n",
    "        #### NOTE: individual Species' sections can be copy-and-pasted (and\n",
    "        #### assigned distinct keys and names), to create additional Species.\n",
    "\n",
    "\n",
    "                }, # <END> 'species'\n",
    "\n",
    "            }, # <END> 'comm'\n",
    "\n",
    "\n",
    "    ##########################################################################\n",
    "\n",
    "    ###############\n",
    "    #### MODEL ####\n",
    "    ###############\n",
    "        'model': {\n",
    "            #total Model runtime (in timesteps)\n",
    "            'T':            600,\n",
    "            #min burn-in runtime (in timesteps)\n",
    "            'burn_T':       50,\n",
    "            #seed number\n",
    "            'num':          None,\n",
    "            #time step interval for simplification of tskit tables\n",
    "            'tskit_simp_interval':      100,\n",
    "\n",
    "            ###############################\n",
    "            #### iterations parameters ####\n",
    "            ###############################\n",
    "            'its': {\n",
    "                #num iterations\n",
    "                'n_its':            1,\n",
    "                #whether to randomize Landscape each iteration\n",
    "                'rand_landscape':   False,\n",
    "                #whether to randomize Community each iteration\n",
    "                'rand_comm':        False,\n",
    "                #whether to burn in each iteration\n",
    "                'repeat_burn':      False,\n",
    "                }, # <END> 'iterations'\n",
    "\n",
    "\n",
    "            ####################################\n",
    "            #### data-collection parameters ####\n",
    "            ####################################\n",
    "            'data': {\n",
    "                'sampling': {\n",
    "                    #sampling scheme {'all', 'random', 'point', 'transect'}\n",
    "                    'scheme':               'transect',\n",
    "                    #sample size at each point, for point & transect sampling\n",
    "                    'n':                    250,\n",
    "                    #coords of collection points, for point sampling\n",
    "                    'points':               None,\n",
    "                    #coords of transect endpoints, for transect sampling\n",
    "                    'transect_endpoints':   [(25, 25), (75, 5)],\n",
    "                    #num points along transect, for transect sampling\n",
    "                    'n_transect_points':    6,\n",
    "                    #collection radius around points, for point & transect\n",
    "                    #sampling\n",
    "                    'radius':               10,\n",
    "                    #when to collect data\n",
    "                    'when':                 100,\n",
    "                    #whether to save current Layers when data is collected\n",
    "                    'include_landscape':    False,\n",
    "                    #whether to include fixed loci in VCF files\n",
    "                    'include_fixed_sites':  True,\n",
    "                    },\n",
    "                'format': {\n",
    "                    #format for genetic data {'vcf', 'fasta'}\n",
    "                    'gen_format':           ['vcf', 'fasta'],\n",
    "                    #format for vector geodata {'csv', 'shapefile', 'geojson'}\n",
    "                    'geo_vect_format':      'csv',\n",
    "                    #format for raster geodata {'geotiff', 'txt'}\n",
    "                    'geo_rast_format':      'geotiff',\n",
    "                    },\n",
    "                }, #<END> 'data'\n",
    "\n",
    "\n",
    "            } # <END> 'model'\n",
    "\n",
    "        } # <END> params\n",
    "\n",
    "    return params\n",
    "\n",
    "\n",
    "def _run(params, save_figs=False, time_it=False, make_gifs=False,\n",
    "         make_3d_plots=False, n_individs_plot=10_000):\n",
    "    params.model['name'] = 'yosemite_demo'\n",
    "    # set the amount of time before and after climate change\n",
    "    t_before_cc = 500\n",
    "    t_after_cc = 100\n",
    "\n",
    "    # create colormap to match the phenotype colors\n",
    "    z_cmap = mpl.cm.coolwarm\n",
    "\n",
    "    # start timer\n",
    "    if time_it:\n",
    "        start = time.time()\n",
    "\n",
    "    # create the model\n",
    "    from .. import make_model\n",
    "    mod = make_model(params, verbose=True)\n",
    "\n",
    "    # set plotting params\n",
    "    ms = 6\n",
    "    ax_fontdict = {'fontsize': 12,\n",
    "                   'name': 'Bitstream Vera Sans'}\n",
    "    ttl_fontdict = {'fontsize': 15,\n",
    "                    'name': 'Bitstream Vera Sans'}\n",
    "\n",
    "    # burn in, then plot starting population, on both rasters\n",
    "    mod.walk(20000, 'burn')\n",
    "\n",
    "    # run for the first 500 timsteps, before climate change\n",
    "    for _ in range(t_before_cc):\n",
    "        mod.walk(1)\n",
    "        if save_figs and make_gifs:\n",
    "            save_model_gif(mod, save_figs, make_gifs,\n",
    "                           n_individs=n_individs_plot)\n",
    "\n",
    "    # calculate neigh-mean raster before climate-change\n",
    "    if with_pykrige:\n",
    "        neigh_mean_b4 = krig_phenotype(mod, save_figs, make_gifs)\n",
    "        try:\n",
    "            np.savetxt('b4_cc_krig.txt', neigh_mean_b4)\n",
    "        except Exception as e:\n",
    "            print(('COULD NOT SAVE KRIGED PHENOTYPE BEFORE CLIMATE CHANGE:'\n",
    "                  ' %s') % e)\n",
    "\n",
    "    # walk for 100 more timesteps, then plot again,\n",
    "    # at end of climate-change period\n",
    "    for _ in range(t_after_cc):\n",
    "        mod.walk(1)\n",
    "        if save_figs and make_gifs:\n",
    "            save_model_gif(mod, save_figs, make_gifs,\n",
    "                           n_individs=n_individs_plot)\n",
    "\n",
    "    # end timer\n",
    "    if time_it:\n",
    "        stop = time.time()\n",
    "        tot_time = stop-start\n",
    "\n",
    "    # calculate neigh-mean raster after climate-change\n",
    "    if with_pykrige:\n",
    "        neigh_mean_af = krig_phenotype(mod)\n",
    "        try:\n",
    "            np.savetxt('af_cc_krig.txt', neigh_mean_af)\n",
    "        except Exception as e:\n",
    "            print(('COULD NOT SAVE KRIGED PHENOTYPE AFTER CLIMATE CHANGE:'\n",
    "                  ' %s') % e)\n",
    "\n",
    "\n",
    "        # calculate the difference between the two neigh-mean rasters\n",
    "        neigh_mean_diff = neigh_mean_af - neigh_mean_b4\n",
    "\n",
    "    # create and save a population-size plot\n",
    "    mod.walk(50)\n",
    "    Nt_fig = plt.figure()\n",
    "    Nt_ax = Nt_fig.add_subplot(111)\n",
    "    #burn_len = mod.burn_t\n",
    "    line_height = int(10000*np.ceil(max(mod.comm[0].Nt[-mod.t:])/10000))\n",
    "    Nt_ax.plot(range(len(mod.comm[0].Nt[-mod.t:])), mod.comm[0].Nt[-mod.t:])\n",
    "    #Nt_ax.plot([burn_len, burn_len], [0, line_height], c='red')\n",
    "    Nt_ax.plot([500, 500], [0, line_height], c='red')\n",
    "    Nt_ax.plot([600, 600], [0, line_height], c='red')\n",
    "    chng_lyr_path = os.path.join(DATA_PATH, 'yosemite_lyrs/ppt/')\n",
    "    chng_yrs = [int(x[:3]) for x in os.listdir(chng_lyr_path)]\n",
    "    for yr in chng_yrs:\n",
    "        Nt_ax.plot([yr, yr], [0, line_height], ':r',\n",
    "                   linewidth=0.5)\n",
    "    Nt_ax.set_label('time (time steps/years)')\n",
    "    Nt_ax.set_ylabel('total population size (individuals)')\n",
    "    if save_figs:\n",
    "        Nt_fig.savefig('yosemite_time_series_pop_size.png', format='png',\n",
    "                       dpi=1000)\n",
    "\n",
    "    # make the change-raster plot\n",
    "    change_fig = make_change_fig(mod)\n",
    "    if save_figs:\n",
    "        change_fig.savefig('yosemite_change_fig.png', format='png', dpi=1000)\n",
    "\n",
    "    # create the GIF using imagemagick\n",
    "    if make_gifs:\n",
    "        try:\n",
    "            os.system('cd %s' % gif_dir)\n",
    "            os.system(('cd %s; convert -delay 5 -loop 0 `ls -v` '\n",
    "                       '../yosemite.gif; cd ..') % gif_dir)\n",
    "            os.system('..')\n",
    "        except Exception as e:\n",
    "            print(('\\nCould not use Imagemagick to create the GIF. '\n",
    "                   'The following error was thrown:\\n\\t%s') % e)\n",
    "\n",
    "    # make the draped-raster plot\n",
    "    if with_pykrige:\n",
    "        DEM = rio.open(os.path.join(DATA_PATH,\n",
    "                                    'yosemite_DEM_90x90.tif')).read()[0, :, :]\n",
    "        if make_3d_plots:\n",
    "            plt.rc('animation', html='html5')\n",
    "            pheno_drape_fig = drape_raster(mod, neigh_mean_b4, DEM, 'phenotype',\n",
    "                                           'yosemite_pheno_drape_fig.gif',\n",
    "                                           save_figs, make_gifs,\n",
    "                                           cmap='coolwarm')\n",
    "        #pheno_chng_cmap = mpl.colors.LinearSegmentedColormap.from_list(\n",
    "        #    'custom', [(0, plt.cm.coolwarm(0.075)), (1,\n",
    "        #                                             plt.cm.coolwarm(0.925))])\n",
    "            change_drape_fig = drape_raster(mod, neigh_mean_diff, DEM,\n",
    "                                            'change in phenotype',\n",
    "                                            'yosemite_pheno_change_drape_fig.gif',\n",
    "                                            save_figs, make_gifs, cmap='PuOr_r')\n",
    "            if save_figs:\n",
    "                pheno_drape_fig.savefig('yosemite_pheno_drape_fig.png',\n",
    "                                        format='png', dpi=1000)\n",
    "                change_drape_fig.savefig('yosemite_pheno_change_drape_fig.png',\n",
    "                                         format='png', dpi=1000)\n",
    "\n",
    "\n",
    "    # print out time\n",
    "    if time_it:\n",
    "        print(\"\\n\\nModel ran in %0.2f seconds.\" % tot_time)\n",
    "\n",
    "    return mod"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python3.7 TF-2.3.0",
   "language": "python",
   "name": "python3.7-tf2.3.0"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
