---
title: "analysis_tests"
output: html_document
---


```{r}
library("tidyverse")
library("viridis")
library("here")
library("ggplot2")
library("lme4")
library("lmerTest")
```

# Pop size
```{r}
df <- read.csv(here("p3_methods", "outputs", "popsize_results.csv"))
df %>% group_by(K) %>% summarize(mean = mean(popsize), min = min(popsize), max = max(popsize))
```

# Phenotype-environment association 
```{r fig.width=10, fig.height=4}
df <- 
  read.csv(here("p3_methods", "outputs", "phenotype_environment_results.csv")) %>%
  filter(sampstrat == "full")

# Calculate the proportion of times the phenotype and environment are correlated
df %>%
  # replace NA with 0, the NA is due to there being no variation
  # (there is one simulation where this occurs for one trait, which has a constant intermediate value of 0.5 across all individuals) 
  # The simulation makes sense: small pop size, weak selection, high migration, low spatial autocorrelation
  # and even then one of the correlations is still significant!
  mutate(cor_sig = case_when(is.na(cor_sig) ~ 0, TRUE ~ cor_sig)) %>%
  # calculate the proportion of times the phenotype ~ environment correlations were significant
  summarize(mean(cor_sig == 1, na.rm = TRUE))
  # result: 99.9% of the time phenotype and environment are correlated

df <- 
  df %>%
    # take the mean of cor1_r and cor2_r
    mutate(cor_r = (cor1_r + cor2_r)/2) %>%
    # replace NA with 0, the NA is due to there being no variation (as described above)
    mutate(cor_r = case_when(is.na(cor_r) ~ 0, TRUE ~ cor_r))

# Run linear model
lmer(cor_r ~ K + m + phi + H + r + (1|seed), data = df) %>% summary()

# Calculate the magnitude of the phenotype-environment correlation across different parameter values
df %>% 
  group_by(H) %>%
  summarize(mean = mean(abs(cor_r), na.rm = TRUE), sd = sd(abs(cor_r), na.rm = TRUE))

df %>% 
  group_by(m) %>%
  summarize(mean = mean(abs(cor_r), na.rm = TRUE), sd = sd(abs(cor_r), na.rm = TRUE))

df %>% 
  group_by(phi) %>%
  summarize(mean = mean(abs(cor_r), na.rm = TRUE), sd = sd(abs(cor_r), na.rm = TRUE))

ggdf <-
  df %>%
  # mutate to numeric
  mutate_at(c("K", "phi", "m", "H", "r"), ~as.numeric(as.character(.))) %>%
  # mutate params to get low and high
    mutate(
      K = case_when(K == min(.[["K"]]) ~ "L", K == max(.[["K"]]) ~ "H"),
      m = case_when(m == min(.[["m"]]) ~ "L", m == max(.[["m"]]) ~ "H"),
      phi = case_when(phi == min(.[["phi"]]) ~ "L", phi == max(.[["phi"]]) ~ "H"),
      H = case_when(H == min(.[["H"]]) ~ "L", H == max(.[["H"]]) ~ "H"),
      r = case_when(r == min(.[["r"]]) ~ "L", r == max(.[["r"]]) ~ "H"),
    ) %>%
  pivot_longer(c(K, m, phi, H, r))

ggplot(ggdf) +
  geom_boxplot(aes(x = value, y = cor_r)) +
  facet_wrap(~name)
```

# Phenotype-environment mismatch
```{r fig.width=10, fig.height=4}
df$eff_mismatch <- df$phi * df$mismatch_mean

ggplot(df[df$sampstrat == "full", ], aes(x = eff_mismatch, fill = factor(phi))) +
  geom_boxplot()

df %>% filter(sampstrat == "full") %>% group_by(phi, K, m, r, H) %>% summarise(mean(abs(mismatch_mean)))
df %>% filter(sampstrat == "full") %>% group_by(phi) %>% summarise(mean(abs(mismatch_mean)))
df %>% filter(sampstrat == "full") %>% group_by(H) %>% summarise(mean(abs(mismatch_mean)))

df_long <-
  df %>%
  filter(sampstrat == "full") %>%
  pivot_longer(c(phi, K, m, H, r)) %>%
  group_by(name) %>%
  mutate(value = case_when(
      value == min(value) ~ "low",
      value == max(value) ~ "high",
      TRUE ~ "NA")) %>%
  mutate(level = factor(value))

ggplot(df_long, aes(y = abs(mismatch_mean), fill = level)) +
  geom_boxplot() +
  facet_wrap(~name, nrow = 1) +
  theme(panel.background = element_rect(fill = NA, color = "gray"),
        strip.background = element_rect(color = "gray"),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank()) +
  ylab("|mean mismatch|")
  
```


# Genotype-environment correlation

```{r}
ge_df <- read.csv(here("p3_methods", "outputs", "genotype_environment_results.csv"))
df <- 
  ge_df %>%
  # note NA values occur because there is no genetic variance, we are treating this as a cor of 0 and a sig (count of significant cors) of 0
  # only a very small amount are NAs (less than 0.01%)
  mutate(cor1_rNA = case_when(is.na(cor1_r) ~ 0, TRUE ~ cor1_r)) %>%
  mutate(cor2_rNA = case_when(is.na(cor2_r) ~ 0, TRUE ~ cor2_r)) %>%
  mutate(cor_sigNA = case_when(is.na(cor_sig) ~ 0, TRUE ~ cor_sig)) 

# calculate proportion of times gea is significant
df %>% 
  filter(adaptive == TRUE) %>%
  summarize(mean((cor_sigNA == 0.5 | cor_sigNA == 1), na.rm = TRUE))

# combine r
ggdf <-
  df %>%
  pivot_longer(c(cor1_rNA, cor2_rNA)) %>%
  mutate(cor_r = abs(value)) 

# run linear model on adaptive loci only
lmer(cor_r ~ K + m + phi + H + r + (1|seed), data = filter(ggdf, adaptive == TRUE)) %>% summary()

ggplot(ggdf) +
  geom_density(aes(fill = adaptive, y = cor_r), alpha = 0.5) +
  facet_wrap(~factor(m))

# no groups
df %>%
  group_by(adaptive) %>%
  summarize(
    cor_r = mean(abs(c(cor1_rNA, cor2_rNA)), na.rm = TRUE), 
    cor_sd = sd(abs(c(cor1_rNA, cor2_rNA)), na.rm = TRUE),
    sig = mean(cor_sigNA, na.rm = TRUE), 
    sd_sig = sd(cor_sigNA, na.rm = TRUE)
    )

# grouped by H
df %>%
  group_by(adaptive, H) %>% 
  # cor_r is the mean absolute value of all correlations
  # cor_sig is the mean number of significant associations at the 0.5 level
  summarize(
    cor_r = mean(abs(c(cor1_rNA, cor2_rNA)), na.rm = TRUE), 
    cor_sd = sd(abs(c(cor1_rNA, cor2_rNA)), na.rm = TRUE),
    sig = mean(cor_sigNA, na.rm = TRUE), 
    sd_sig = sd(cor_sigNA, na.rm = TRUE)
    )

# grouped by H and gene flow
df %>%
  group_by(adaptive, H, m) %>% 
  # cor_r is the mean absolute value of all correlations
  # cor_sig is the mean number of significant associations at the 0.5 level
  summarize(
    cor_r = mean(abs(c(cor1_rNA, cor2_rNA)), na.rm = TRUE), 
    cor_sd = sd(abs(c(cor1_rNA, cor2_rNA)), na.rm = TRUE),
    sig = mean(cor_sigNA, na.rm = TRUE), 
    sd_sig = sd(cor_sigNA, na.rm = TRUE)
    )
```

```{r}
# calculate the proportion of times the neutral snps have a greater correlation with the environment than the adaptive snps
calc_empp <- function(sub_df){
  env1_adaptive <- sub_df[1:4, "cor1_r"]
  env2_adaptive <- sub_df[5:8, "cor2_r"]
  env1_neutral <- sub_df[-c(1:8), "cor1_r"]
  env2_neutral <- sub_df[-c(1:8), "cor2_r"] 

  env1_empp <- mean(unlist(map(env1_adaptive, ~abs(.x) <= abs(env1_neutral))), na.rm = TRUE)
  env2_empp <- mean(unlist(map(env2_adaptive, ~abs(.x) <= abs(env1_neutral))), na.rm = TRUE)
  
  ls <-
    list(
      m = unique(sub_df$m),
      H = unique(sub_df$H),
      r = unique(sub_df$r),
      K = unique(sub_df$K),
      phi = unique(sub_df$phi),
      seed = unique(sub_df$seed),
      env1_empp = env1_empp, 
      env2_empp = env2_empp, 
      env_empp = mean(c(env1_empp, env2_empp))
      )

  return(ls)
}

df_split <- df %>% group_by(K, phi, m, seed, H, r, it) %>% group_split()
empp <- map(df_split, ~calc_empp(.x), .progress = TRUE)
empp_df <- empp %>% bind_rows()

# run linear model
lmer(env_empp ~ K + m + phi + H + r + (1|seed), data = empp_df) %>% summary()

# grouped by gene flow
empp_df %>% 
  group_by(m) %>% 
  summarize(
    mean = mean(env_empp, na.rm = TRUE),
    sd = sd(env_empp, na.rm = TRUE)
    )

# grouped by H
empp_df %>% 
  group_by(H) %>% 
  summarize(
    mean = mean(env_empp, na.rm = TRUE),
    sd = sd(env_empp, na.rm = TRUE)
    )

# grouped by H and m
empp_df %>% 
  group_by(H, m) %>% 
  summarize(
    mean = mean(env_empp, na.rm = TRUE),
    sd = sd(env_empp, na.rm = TRUE)
    )
```
